### ISSUE TITLE
Improve code coverage for System.Linq

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#1143]
According to the current code coverage report, this library is only at ~24%. Our general goal is closer to 90%, though this will depend on the library. We should investigate to see whether the current coverage level is appropriate, and if not, improve our test suites to provide better coverage.

**Comment by rafidka** [comment]
@stephentoub, I am interested in working on this task, but I think I need some more explanation about what I need to do. For example, do I need to go over all methods and see which ones don't have unit tests? If you just give me some guideline then I can take a look and get back to you on whether I can do it or not.

**Comment by stephentoub** [comment]
@rafidka, thanks for your interest.  I've put up some information on a [[Code Coverage]] wiki page.  Let us know if that doesn't address your questions.

**Comment by stephentoub** [comment]
@AlekseyTs, are more tests forthcoming for System.Linq, or is what's up there now the full available set?

**Comment by Clockwork-Muse** [comment]
@rafidka 
I've been working on the Linq.Parallel tests, would any of that be applicable/transferrable/helpful?  I've got tests for all the individual operators (`All`, `Where`, `Select`, etc)... although I think I need to "label" sources so it's easier to tell which one is failing (ie, array versus enumerable input source).

From the look of what's out there (System.Linq/tests), you're going to need to actually _write_ all the tests.  The repository doesn't appear to have anything really useful (it's checking with an empty source only, and expecting success states only).

**Comment by rafidka** [comment]
Thanks for your reply @stephentoub. I just wanted to have a look at the source code, but I couldn't even find the System.Linq project! Or am I missing something?

**Comment by rafidka** [comment]
@Clockwork-Muse yeah, I am sure it will be beneficial to have a look at them. Thanks!

**Comment by Maxwe11** [comment]
@rafidka https://github.com/dotnet/corefx/tree/master/src/System.Linq

**Comment by rafidka** [comment]
Yeah, sorry my bad, I realized I forgot to sync. OK then, I will have a look at it today.

**Comment by rafidka** [comment]
@stephentoub I had a look at the code and I think I can do it. However, do we have a deadline here for getting this done? I don't think this will need months definitely, but just wanted to make sure there is no tight deadline for it.

Also, I noticed that System.Linq lacks documentation and it needs some refactoring (everything is in one big file of around 3000 lines, additionally ReSharper shows 178 warnings in Enumerable.cs). Should I add a task for documentation and another for refactoring?

**Comment by stephentoub** [comment]
> I had a look at the code and I think I can do it

Great. Thanks!

> I noticed that System.Linq lacks documentation

You mean XML documentation?  The MSDN docs for Enumerable are at https://msdn.microsoft.com/en-us/library/vstudio/system.linq.enumerable.aspx .  We already plan to update / augment the XML docs in the code across all of CoreFX automatically based on the MSDN docs (@davkean, FYI).  We should not do so here manually.

> and it needs some refactoring

Feel free to open a separate issue for that.  Please don't do so as part of efforts to improve the testing coverage for the library.

**Comment by rafidka** [comment]
Great, can you please assign this task to myself then? I can't do it myself, as I am not a collaborator on corefx yet.

**Comment by AlekseyTs** [comment]
@stephentoub Regarding "are more tests forthcoming for System.Linq" question. That is all we have ready for open source at the moment.

**Comment by stephentoub** [comment]
@AlekseyTs, thanks.   Just to verify, though, does that mean we don't have any more tests, or does that mean we do have more tests written that just aren't ready to be open sourced but will be at some point?  I'm asking simply because I don't want @rafidka to do duplicative work... if we have more tests that will eventually be released, then his energy would be better spent elsewhere.

**Comment by rafidka** [comment]
@stephentoub, I understand that I have to wait for a reply from @AlekseyTs before working on this task, right?

**Comment by stephentoub** [comment]
@rafidka, thanks. Hold off for now; I want to investigate a little more... I think there are more tests, and I'd like to avoid you doing duplicate work, if possible. I'll get back to you soon.

**Comment by rafidka** [comment]
Thanks @stephentoub for letting me know. I will put #1149 on hold as there is not much point in refactoring before we get the full code.

**Comment by stephentoub** [comment]
@rafidka, I did some digging around, and it looks like we do have tests, but they need a lot of work, so much that it'd probably be just as fast to write new ones from scratch rather than to port them.  So, feel free to go ahead and work on this.  If you also plan to do something around #1149, I'd ask that you do this first; refactoring can introduce bugs, and it'd be important to have tests to catch any such regressions. (I'm also not sure we actually want to see a lot of refactoring done in the actual product code... worth a discussion on #1149.) Thanks!

**Comment by rafidka** [comment]
OK, thanks for the update, I will start on this task soon this. As for #1149, I will add my comments there.

**Comment by rafidka** [comment]
@stephentoub, I am trying to run the unit tests but I am not being able to; I am getting this error when I try to run the tests using ReSharper:

Unable to run xUnit.net tests: Unknown test framework: Could not find xunit.dll or xunit.execution.desktop.dll.
Parameter name: C:\Users\Rafid\Workspace\corefx\bin\Windows_NT.AnyCPU.Debug\System.Linq.Tests\System.Linq.Tests.dll

When I check the path above, indeed I don't see the specified dlls. Notice that I built the whole coreclr from the command line (using build.cmd).

I also tried msbuild /t:BuildAndTest but it doesn't seem to be doing much, as even when I put some invalid code that shouldn't compile, it still passes.

Any idea?

**Comment by stephentoub** [comment]
> I am trying to run the unit tests but I am not being able to

You can't use the desktop xunit, which presumably is what ReSharper is trying to use.  It needs to be run with .NET Core.  If you look in the msbuild output, you can see the command-line that's used to invoke the tests.

> I also tried msbuild /t:BuildAndTest

From what directory?  You need to do it in the System.Linq\tests directory.  If that's where you're doing it, what output are you seeing?

**Comment by rafidka** [comment]
Ah, yeah, it worked now when I tried msbuild /t:BuildAndTest from the tests directory. Just wondering, is there a better to execute the tests than the command line? It is a little bit hard to follow it and see what the problem is.

**Comment by Maxwe11** [comment]
>Just wondering, is there a better to execute the tests than the command line?

in VS2015 only
https://github.com/dotnet/corefx/wiki/Developer-Guide

**Comment by rafidka** [comment]
Oh yeah, I should have remembered that. I tried it now and it is working. Thanks @Maxwe11 .

**Comment by rafidka** [comment]
I added some simple tests and sent a PR to get some feedback. Please let me know if you have any concern, otherwise I will proceed with adding more tests.

One thing I would like to discuss is the input we should feed to the tests. Currently, I am using Enumerable.Range. However, I don't think this is realistic, and I believe we should use some randomly generated data. Obviously, for our tests to be deterministic and so that we can debug them if they fail, we can't make our unit tests use different inputs in different runs. So how about we generate some random enumerables and store them in some files and use them in the tests?

**Comment by Clockwork-Muse** [comment]
... for this kind of testing, isn't `Enumerable.Range(...)` _already_ a custom enumerator?

You can do something akin to what I'm doing - generating via `Enumerable.Range(...).AsParallel()`, ParallelEnumerable.Range(...)`, `sourceArray.AsParallel()`, `Paritioner.Create(sourceArray).AsParallel()` (Obviously you wouldn't do it with parallel versions...).  Maybe add `Iterator`?  Note that Linq isn't going to have any idea about the contents of the array-sourced stuff, so that's nearly as good as random.

I didn't really consider creating a custom `ParallelQuery<T>` type... and I'm not sure how much the testing would be about if Linq was working, or that the custom enumerator was.  That is, all the stuff I mentioned previously ends up being the reference implementation for anything else.

Probably the only thing you want/need random (or pre-shuffled) data for would be the ordering operators (ie `OrderBy`), where it's more to evaluate that the algorithm can handle dealing with the shuffled items.  Just about everything else takes a func to pick out keys for stuff anyways - it has no way to tell what the modification is until it's called on the element.

**Comment by rafidka** [comment]
Thanks for the comments everyone. I made some changes to improve the code. Could you please have a look now and let me know if it looks good, so that I can proceed with adding more tests.

**Comment by Maxwe11** [comment]
@rafidka you doing something wrong because 31c4df0 not displayed in your PR #1175.

**Comment by rafidka** [comment]
@Maxwe11, I think the reason is that I created a new branch for the new change, because I forgot to do so initially so I was working against my master branch. What do you think I should do?

**Comment by Maxwe11** [comment]
@rafidka continue working with `master` branch or close #1175 and open new one

**Comment by rafidka** [comment]
OK, I updated my master branch and submitted it so that the PR is updated. Once it is approved and merged, I will use the new branch and create a new PR.
